### ISSUE TITLE
hasOne association with non primary key

### ISSUE THREAD

**Issue by tzachristas** [sequelize/sequelize#8105]
Hello, 
We are having to models with hasOne association on a non-PrimaryKey column

modelA {id(PrimaryKey), someID, propertyA} And 

modelB{id(PrimaryKey), PropertyA}

modelA hasOne with modelB on modelA.someID = modelB.id

It seems that there is no way yo properly create that association with targetKey and Belongsto.

**Comment by p0vidl0** [comment]
Its not working?
```js
modelA.hasOne(modelB, {foreignKey: 'id', sourceKey: 'someID'});
modelB.belongsTo(modelA, {foreignKey: 'id', targetKey: 'someID'});
```

**Comment by tzachristas** [comment]
It is not properly working. It associates modelB.ID with modelA Primary key and NOT with modelA sourceKey.

SELECT * 
FROM modelA LEFT OUTER JOIN modelB ON modelA.`ID` = modelB.`ID`;

**Comment by p0vidl0** [comment]
can you provide [more detailed](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md#tell-us) information please?

**Comment by tzachristas** [comment]
Sequelize version 4.4.8
Dialect Mysql

  var modelA = sequelize.define("TABLE1", {
    ID: {
      type: DataTypes.INTEGER(11),
      primaryKey: true
    },
    PROPERTY_ID: {
      type: DataTypes.INTEGER(11),
      allowNull: false
    }
    
  modelA.associate = function(models) {
    models.TABLE1.hasOne(models.TABLE2, {foreignKey: 'ID', sourceKey: 'PROPERTY_ID'})
  }


============================================================
  var modelB = sequelize.define("TABLE2", {
    ID: {
      type: DataTypes.INTEGER(11),
      primaryKey: true
    },
    NAME: {
      type: DataTypes.STRING
      allowNull: false
    }
    
  modelB.associate = function(models) {
    models.TABLE2.belongsTo(models.TABLE1, {foreignKey: 'ID', targetKey: 'PROPERTY_ID'})
  }

===========================================================

models.TABLE1.findAll({include: [models.TABLE2]}).then(o => {
      res.send(o);
 })

It create an SQL 
SELECT *
FROM modelA LEFT OUTER JOIN modelB ON modelA.ID = modelB.ID;

and NOT

SELECT *
FROM modelA LEFT OUTER JOIN modelB ON modelA.PROPERTY_ID = modelB.ID;

**Comment by p0vidl0** [comment]
imo: you had invert relations
think its must be like:
```js
models.TABLE1.belongsTo(models.TABLE2, {foreignKey: 'PROPERTY_ID', targetKey: 'ID'})
models.TABLE2.hasOne(models.TABLE1, {foreignKey: 'PROPERTY_ID', sourceKey: 'ID'})
```

**Comment by tzachristas** [comment]
Yes it is working that way...but reverting the association the hasOne must become hasMany.

TABLE1 hasOne TABLE2 but TABLE2 hasManyTABLE1

**Comment by nmehta01** [comment]
I agree with @tzachristas and @dsjstapler that hasOne should support sourceKey

**Comment by erk3** [comment]
+1

**Comment by mattfysh** [comment]
Just ran into this issue... having to use `hasMany` for now instead of `hasOne`... and then validate length of associated model array is zero or one

**Comment by fmhc** [comment]
+1 for sourceKey support in hasOne!

**Comment by afreeland** [comment]
+1 Glad to see others stumbled into this...

**Comment by sureshHARDIYA** [comment]
+1 for sourceKey support in hasOne!

**Comment by dallasclark** [comment]
+1 for sourceKey support in hasOne please

**Comment by bhutten** [comment]
+1 for sourceKey support in hasOne please

**Comment by lessu** [comment]
+1 for sourceKey support in hasOne please

**Comment by redexp** [comment]
Her my temp workaround for `v4.37.5`
```javascript
Table1.hasOne(Table2, {foreignKey: 'table_2_column_id'});
var item = Table1.associations[Table2.name];
item.sourceIdentifier = item.sourceKey = item.sourceKeyAttribute = 'table_1_column_id';
item.sourceKeyIsPrimary = false;
```

**Comment by sushantdhiman** [comment]
hasOne now supports `sourceKey` https://github.com/sequelize/sequelize/pull/9382 , It is available in `5.0.0-beta.5` and above

**Comment by eklenin** [comment]
> Here my temp workaround for `v4.37.5`

Does anybody know similar hack for sequelize v3?
